submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/service/DoiService.kt

```diff
@@ -0,0 +1,125 @@
+package ac.uk.ebi.biostd.submission.service
+
+import ac.uk.ebi.biostd.common.properties.DoiProperties
+import ac.uk.ebi.biostd.submission.exceptions.InvalidAuthorAffiliationException
+import ac.uk.ebi.biostd.submission.exceptions.InvalidAuthorNameException
+import ac.uk.ebi.biostd.submission.exceptions.InvalidDoiException
+import ac.uk.ebi.biostd.submission.exceptions.InvalidOrgException
+import ac.uk.ebi.biostd.submission.exceptions.InvalidOrgNamesException
+import ac.uk.ebi.biostd.submission.exceptions.MissingAuthorAffiliationException
+import ac.uk.ebi.biostd.submission.exceptions.MissingDoiFieldException
+import ac.uk.ebi.biostd.submission.exceptions.MissingTitleException
+import ac.uk.ebi.biostd.submission.model.Contributor
+import ac.uk.ebi.biostd.submission.model.DoiRequest
+import ac.uk.ebi.biostd.submission.model.SubmitRequest
+import ebi.ac.uk.commons.http.ext.RequestParams
+import ebi.ac.uk.commons.http.ext.post
+import ebi.ac.uk.io.FileUtils
+import ebi.ac.uk.model.Section
+import ebi.ac.uk.model.Submission
+import ebi.ac.uk.model.constants.SubFields.DOI
+import ebi.ac.uk.model.extensions.allSections
+import ebi.ac.uk.model.extensions.title
+import ebi.ac.uk.util.collections.ifNotEmpty
+import mu.KotlinLogging
+import org.springframework.core.io.FileSystemResource
+import org.springframework.http.HttpHeaders
+import org.springframework.http.MediaType.MULTIPART_FORM_DATA
+import org.springframework.util.LinkedMultiValueMap
+import org.springframework.web.reactive.function.client.WebClient
+import java.nio.file.Files
+
+private val logger = KotlinLogging.logger {}
+
+@Suppress("ThrowsCount", "ReturnCount")
+class DoiService(
+    private val webClient: WebClient,
+    private val properties: DoiProperties,
+) {
+    fun calculateDoi(accNo: String, rqt: SubmitRequest): String? {
+        val doi = rqt.submission.attributes.find { it.name == DOI.name } ?: return null
+        val previousDoi = rqt.previousVersion?.doi
+
+        if (previousDoi != null) {
+            require(doi.value == previousDoi) { throw InvalidDoiException() }
```

Comment:
I would include in the exception the previous and current value so error is easier to identify.